persist/Persist.js

```diff
@@ -0,0 +1,97 @@
+/*
+ * This file belongs to Hoist, an application development toolkit
+ * developed by Extremely Heavy Industries (www.xh.io | info@xh.io)
+ *
+ * Copyright © 2020 Extremely Heavy Industries Inc.
+ */
+
+import {cloneDeep, isUndefined} from 'lodash';
+import {PersistenceProvider} from './provider/PersistenceProvider';
+
+/**
+ * Decorator to make a class property persistent, syncing its value via a configured
+ * `PersistenceProvider` to maintain and restore mutable values across browser sessions.
+ *
+ * This decorator may be used on any @observable or @bindable class property which is a primitive.
+ * It will initialize the observable's value from the class's default `PersistenceProvider` and will
+ * subsequently write back any changes to the property to that Provider. If the Provider has not
+ * yet been populated with a value, or an error occurs, it will use the value set in-code instead.
+ *
+ * This decorator should always be applied "before" the mobx decorators, i.e. closest to definition,
+ * second in file line order: `@bindable @persist fooBarFlag = true`
```

Comment:
Curious... what happens if this ordering is not adhered to? Does it break in an obvious way, or an "invisible" way?

I expect this to be a common source of bugs if the break is quite subtle. If the break is hard to spot, could we somehow detect and throw?